codesign/codesign.go

```diff
@@ -0,0 +1,338 @@
+		}
+	}
+
+	return m.project, nil
+}
+
+// PrepareCodesigning ...
+func (m *Manager) PrepareCodesigning() (Result, error) {
+	if m.opts.AuthType == NoAuth {
+		m.logger.Println()
+		m.logger.Infof("Skip downloading any Code Signing assets")
+
+		return Result{}, nil
+	}
+
+	strategy, reason, err := m.selectCodeSigningStrategy(m.appleAuthCredentials)
+	if err != nil {
+		m.logger.Warnf("%s", err)
+	}
+
+	switch strategy {
+	case codeSigningXcode:
+		{
+			m.logger.Println()
+			m.logger.Infof("Preparing for Xcode-managed code-signing")
+			m.logger.Printf(reason)
+			m.logger.Println()
+			m.logger.Infof("Downloading certificates from Bitrise")
+			if err := m.downloadAndInstallCertificates(); err != nil {
+				return Result{}, err
+			}
+
+			needsTestDevices := autocodesign.DistributionTypeRequiresDeviceList([]autocodesign.DistributionType{m.opts.ExportMethod})
+			if needsTestDevices && m.opts.RegisterTestDevices && m.bitriseConnection != nil && len(m.bitriseConnection.TestDevices) != 0 {
+				if err := m.registerTestDevices(m.appleAuthCredentials, m.bitriseConnection.TestDevices); err != nil {
+					return Result{}, err
+				}
+			}
+
+			return Result{
+				XcodebuildAuthParams: m.appleAuthCredentials.APIKey,
+			}, nil
+		}
+	case codeSigningBitriseAPIKey, codeSigningBitriseAppleID:
+		{
+			m.logger.Println()
+			m.logger.Infof("Bitrise-managed code-signing")
+			m.logger.Printf(reason)
+			if err := m.prepareCodeSigningWithBitrise(m.appleAuthCredentials); err != nil {
+				return Result{}, err
+			}
+
+			return Result{}, nil
+		}
+	}
+
+	return Result{}, nil
+}
+
+// SelectConnectionCredentials ...
+func SelectConnectionCredentials(authType AuthType, conn *devportalservice.AppleDeveloperConnection, logger log.Logger) (appleauth.Credentials, error) {
+	var authSource appleauth.Source
+
+	switch authType {
+	case APIKeyAuth:
+		authSource = &appleauth.ConnectionAPIKeySource{}
+	case AppleIDAuth:
+		authSource = &appleauth.ConnectionAppleIDFastlaneSource{}
+	case NoAuth:
+		panic("not supported")
+	default:
+		panic("missing implementation")
+	}
+
+	authConfig, err := appleauth.Select(conn, []appleauth.Source{authSource}, appleauth.Inputs{})
+	if err != nil {
+		if conn != nil && conn.APIKeyConnection == nil && conn.AppleIDConnection == nil {
+			fmt.Println()
+			logger.Warnf("%s", devportalclient.NotConnectedWarning)
+		}
+
+		if errors.Is(err, &appleauth.MissingAuthConfigError{}) {
+			if authType == AppleIDAuth {
+				return appleauth.Credentials{}, fmt.Errorf("Apple ID authentication is selected in Step inputs, but Bitrise Apple service connection is unset")
+			}
+
+			return appleauth.Credentials{}, fmt.Errorf("API key authentication is selected in Step inputs, but Bitrise Apple service connection is unset")
+		}
+
+		return appleauth.Credentials{}, fmt.Errorf("could not select Apple authentication credentials: %w", err)
+	}
+
+	if authConfig.APIKey != nil {
+		authConfig.AppleID = nil
```

Comment:
for the sake of clarity should not we clear `authConfig.APIKey` if `authConfig.AppleID != nil` ?

Reply:
For consistecy yes, should not really matter either way.